### PR TITLE
Master+mbed linux smc

### PR DESCRIPTION
**Pull request by bryanodonoghue**
Hi.

This set of patches enables a new SMC call designed to allow OPTEE to populate an in-memory DTB which is loaded after OPTEE has handed control back to NS-world.

Tested on WaRP7 with an i.MX7S

### CODE REVIEW

File: core/arch/arm/plat-imx/sm_platform_handler.c [hunk 1]
```diff
@@ -41,6 +43,38 @@ static bool imx_sip_handler(struct thread_smc_args *smc_args)
 	return false;
 }
 
+#ifdef CFG_SM_PLATFORM_MBED_LINUX
+static void imx_oem_populate_fdt(struct thread_smc_args *smc_args)
+{
+	/* Populate with generic_boot provided parameters */
+	init_fdt(smc_args->a1);
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-imx/sm_platform_handler.c]: This address need to be checked. What if it's a secure address?

Comment (bryanodonoghue) [reply core/arch/arm/plat-imx/sm_platform_handler.c]: I thought of that and tested malicious input.
It behaves the same way as giving an invalid DTB pointer as an OPTEE arg.

=> smc 0x83000001 0
E/TC:0   init_fdt:786 Device Tree missing
Res:  0 0 0 0

**Comment by MrVan** [comment]
@bryanodonoghue So this is to support SPL -> OP-TEE->U-Boot->Linux and SPL->OP-TEE->Linux?
If this is the case, the call should only be called once and close the gate.

**Comment by jenswi-linaro** [comment]
Actually I think this is a very dangerous way of doing it, which I doubt can ever be made completely secure, the attack surface is simply too large.
We should consider OP-TEE providing a DTB overlay instead.
There's some examples at https://www.raspberrypi.org/documentation/configuration/device-tree.md

**Comment by bryanodonoghue** [comment]
I can look into injecting a DT overlay into some sort of known location that OPTEE/u-boot and kernel agree on..

**Comment by bryanodonoghue** [comment]
@MrVan 
No we are looking to-do
ATF -> FIP {OPTEE, u-boot}
OPTEE
u-boot
u-boot -> FIT {kernel, initramfs, kernel-dtb}

In that case the DTB will not be passed by ATF (or u-boot) as an OPTEE argument.

My first thought was to make an SMC call so that after uboot had loaded the DTB - it could call into OPTEE and have OPTEE populate.

@jenswi-linaro has suggested a DTB overlay, which now that I think on it might be a highly preferable way of doing it.

OPTEE would place a DTB overlay at a known location in memory. u-boot or kernel then could consume that in-memory binary blob.

In principle, I think that can work.

**Comment by bryanodonoghue** [comment]
@jenswi-linaro 
I see you did the original libfdt import based on v1.4.1 - the first option for picking up the overlay code is 1.4.2 but, the latest version is 1.4.7 - including a few fixes along the way.
Unless you think 1.4.2 is the right target, I'll aim to import 1.4.7 as a first step to adding overlay functionality here.

**Comment by jenswi-linaro** [comment]
1.4.7 sounds good

**Comment by MrVan** [comment]
+1 DTB overlay.

**Comment by bryanodonoghue** [comment]
First PR as a result of this discussion here: https://github.com/OP-TEE/optee_os/pull/2647

Closing this one

**Pull request closed without merging**